File: bench/lang/method_call.rb
class Toggle
  attr_reader :value

  def initialize(start_state)
    @value = start_state
  end

  def activate
    @value = !@value
    self
  end
end

class NthToggle < Toggle
  def initialize(start_state, max_counter)
    super(start_state)
    @count_max = max_counter
    @count = 0
  end

  def activate
    @count += 1
    if @count >= @count_max
      @value = !@value
      @count = 0
    end
    self
  end
end

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)

n = 100000
val = true
toggle = Toggle.new(val)

n.times do
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
  val = toggle.activate.value
end

ntoggle = NthToggle.new(val, 3)

n.times do
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
  val = ntoggle.activate.value
end

elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
puts "elapsed: #{elapsed}"

File: bench/lang/method_call.py
import time

class Toggle:
    def __init__(self, start_state):
        self.state = start_state

    @property
    def value(self):
        return self.state

    def activate(self):
        self.state = not self.state
        return self

class NthToggle(Toggle):
    def __init__(self, start_state, max_counter):
        super().__init__(start_state)
        self.count_max = max_counter
        self.count = 0

    def activate(self):
        self.count += 1
        if self.count >= self.count_max:
            self.state = not self.state
            self.count = 0
        return self

start = time.time()

n = 100000
val = True
toggle = Toggle(val)

for _ in range(n):
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value
    val = toggle.activate().value

ntoggle = NthToggle(val, 3)

for _ in range(n):
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value
    val = ntoggle.activate().value

elapsed = time.time() - start
print("elapsed: %s" % elapsed)
